### PR TITLE
10635 delete layers on user deletion

### LINKED ISSUE #10635: Deleting an user doesn't delete its layers
When deleting an user, its layers do not get deleted:
 - In `User.before_destroy` the maps are destroyed, but that doesn't destroy the layers. Deleting a visualization works (because there is layer deletion on visualization destruction)
 - Custom basemaps (user layers) are disassociated but not deleted.

cc @CartoDB/builder-backend

### PR DESCRIPTION
**Pull request by javitonino**
Fixes #10635 and #10792 (by deleting visualizations before the user)

The trick is in `UserTable.after_destroy`. In this case, `super` triggers map destruction (via Sequel "FK"), and `service.after_destroy` triggers visualization destruction. If the map is destroyed first, the reference to the layers is lost.

~~Maybe layers should be destroyed from map (instead of from visualization), but I'm feeling cautious and I think this could be done as part of the move to AR model.~~
Also, layers are now destroyed from the map instead of the visualization. This forced to do some changed to how named maps are handled during deletions.

### CODE REVIEW

**Comment by javitonino** [comment]
CR @gfiorav

**Comment by javitonino** [comment]
ping!

File: spec/models/user_spec.rb [hunk 1]
```diff
@@ -1161,25 +1161,24 @@ def overquota(delta = 0)
...
-      .returns(true)
+                    .with(".*#{uuid}:vizjson")
+                    .at_least_once
+                    .returns(true)
 
     doomed_user.destroy
 
-    DataImport.where(:user_id => doomed_user.id).count.should == 0
-    UserTable.where(:user_id => doomed_user.id).count.should == 0
+    DataImport.where(user_id: doomed_user.id).count.should == 0
```
Comment (gfiorav) [inline spec/models/user_spec.rb]: is this better?

File: spec/models/user_spec.rb [hunk 2]
```diff
@@ -2320,35 +2320,52 @@ def stub_and_check_version_pre_mu(version, is_pre_mu)
...
+      visualization = CartoDB::Visualization::Member.new(id: carto_visualization.id).fetch
+
+      # Force ORM to cache layers (to check if they are deleted later)
+      canonical_visualization.map.layers
+      visualization.map.layers
+
+      user_layer = Layer.create(kind: 'tiled')
+      user.add_layer(user_layer)
+
+      return user, table, [canonical_visualization, visualization], user_layer
```
Comment (gfiorav) [inline spec/models/user_spec.rb]: +1

Comment (javitonino) [inline spec/models/user_spec.rb]: It is not redundant here, it is needed for multi-return. Just removing the return gives an error. Maybe there is another way to do it?

File: spec/models/user_spec.rb (hunk 1 above)
Comment (javitonino) [inline spec/models/user_spec.rb]: This doesn't work with `should` :(

It seems it be used for normal comparisons though:
http://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Style/NumericPredicate

File: spec/models/user_spec.rb (hunk 2 above)
Comment (gfiorav) [inline spec/models/user_spec.rb]: yeah! you have to:

```ruby
[user, table, [canonical_visualization, visualization], user_layer]
```

sorry I didn't realise in the first one

Comment (javitonino) [inline spec/models/user_spec.rb]: You learn something new everyday. I thought it was special syntax, not just an array 🌴

**Comment by javitonino** [comment]
This breaks visualization deletion, cannot believe the tests didn't catch it 😞

**Comment by javitonino** [comment]
Ready for a second CR round @gfiorav. Please note that this is merged with #10837 so I can have proper tests, so a lot of the changes are being reviewed separately. If you prefer, we can wait until the other PR is in master so the changes of this PR are easier to read.

**Comment by javitonino** [comment]
Merged with master

**Pull request merged**